Add item to list and select frequency of purchase

edenbekele opened the pull request:
_For an example of how to fill this template out, [see this Pull Request](https://github.com/the-collab-lab/tcl-3-smart-shopping-list/pull/44)._ 

## Description

- Users can add items to a list and select purchase frequency 
- When a user adds an item to the list, the Firestore database updates the list associated with the users token
- Display items from the list in the list view on our application
- Installed `react-hooks-form` dependency 
- Added validations and error handling


## Related Issue

Issue #4 

## Acceptance Criteria

-User is presented with a form that allows them to enter the name of the item and select how soon they anticipate needing to buy it again (Soon, Kind of Soon, Not Soon)
-When the user submits the form, the item is saved to the database, associated with the user’s token
-Along with the item name, an integer corresponding to the estimated number of days until next purchase is saved: 7 for “Soon”, 14 for “Kind of Soon”, and 30 for “Not Soon”
-The last purchased date should be set to null initially (i.e. you can create an item without purchasing it)
-Item names should be displayed as a simple, unordered list on the list view after they’re added

## Type of Changes

<!-- Put an `✓` for the applicable box: -->

|     | Type                       |
| --- | -------------------------- |
|    | :bug: Bug fix              |
|  ✓ | :sparkles: New feature     |
|    | :hammer: Refactoring       |
|    | :100: Add tests            |
|  ✓   | :link: Update dependencies |
|    | :scroll: Docs              |

## Updates

### Before

![tcl-before](https://user-images.githubusercontent.com/61370161/96019858-c1e2d280-0e12-11eb-97c9-49b3368ac008.png)


### After

![Screen Shot 2020-10-14 at 11 46 10 AM](https://user-images.githubusercontent.com/61370161/96020041-15edb700-0e13-11eb-8d47-5bb36f620872.png)

![Screen Shot 2020-10-14 at 11 46 25 AM](https://user-images.githubusercontent.com/61370161/96020071-1c7c2e80-0e13-11eb-8ce0-d7a23e77098e.png)

![Screen Shot 2020-10-14 at 11 47 14 AM](https://user-images.githubusercontent.com/61370161/96020082-21d97900-0e13-11eb-8014-1144c66c1459.png)


## Testing Steps / QA Criteria

- Go to the branch (`git checkout tc-eb-add-item-to-list`) and pull it down (`git pull origin tc-eb-add-item-to-list`)
- Run `npm install`
- Open Firestore database and go to the lists collection and the test-token document
- Go to "Add an Item". Add item name and select frequency then submit. Database and list view should automatically update. You should receive an alert when item is submitted.
- Make a change in the Firestore database to view real time updates on the application
- Try to submit a blank item to view validation error

edenbekele committed:
Added form for user to add an item to a list

Co-authored-by: Eden Bekele <edbek@yahoo.com>
Co-authored-by: Terry Chiem <terrychiem@gmail.com>

edenbekele committed:
Connected form to Firestore to add list items

Co-authored-by: Eden Bekele <edbek@yahoo.com>
Co-authored-by: Terry Chiem <terrychiem@gmail.com>

terchiem committed:
Changed firestore collection for items into an array.

Co-authored-by: Eden Bekele <edbek@yahoo.com>
Co-authored-by: Terry Chiem <terrychiem@gmail.com>

terchiem committed:
Implement loading of firestore list and displaying of items in an unordered list.

Co-authored-by: Eden Bekele <edbek@yahoo.com>
Co-authored-by: Terry Chiem <terrychiem@gmail.com>

terchiem committed:
Converted List to use React Firestore components for real time updates.

Co-authored-by: Eden Bekele <edbek@yahoo.com>
Co-authored-by: Terry Chiem <terrychiem@gmail.com>

edenbekele committed:
Converted form into a React hook form and added validations and error handling

Co-authored-by: Terry Chiem <terrychiem@gmail.com
Co-authored-by: Eden Bekele <edbek@yahoo.com>

edenbekele committed:
Added feedback for when an item is added

Co-authored-by: Terry Chiem <terrychiem@gmail.com
Co-authored-by: Eden Bekele <edbek@yahoo.com>

alexandra-lim commented on src/Components/AddItem.jsx:
```diff
@@ -1,5 +1,76 @@
 export default function AddItem() {
-  return <div>Hi from Add an Item View!</div>;
+  const [token, setToken] = useState('test-token');
+  const { register, handleSubmit, watch, errors } = useForm();
```
This isn't a blocker, but we noticed `watch` wasn't being used in the code. Is this something we can remove?

alexandra-lim commented on src/Components/AddItem.jsx:
```diff
@@ -1,5 +1,76 @@
+
+  return (
+    <form onSubmit={handleSubmit(onSubmit)}>
+      <br />
```
Also not a blocker, but are these `<br>` tags temporary styling for this issue?

edenbekele replied:
Yes, the breaks were used for temporary styling so that the app looked a little better :)

GabbyJ commented on src/Components/AddItem.jsx:
```diff
@@ -1,5 +1,76 @@
+          lastPurchased: null,
+        }),
+      })
+      .then(() => alert('Item has been submitted!'));
```
This looks great! We think it would be good to have the data reset after a successful submission. [This link](https://react-hook-form.com/api/#reset) looks similar to the current form code and may be helpful to get that working.

edenbekele replied:
Thanks for the suggestion! We just added this functionality.

GabbyJ replied:
Everything looks great now. Good job!

alexandra-lim commented on src/Components/AddItem.jsx:
```diff
@@ -1,5 +1,76 @@
+  };
+
+  return (
+    <form onSubmit={handleSubmit(onSubmit)}>
```
This form looks great! I haven't used React Hooks Form before, it's super neat! This is more of an extra credit (aka not a blocker), but it'd be cool to leverage some of the accessibility features they have to make our app more usable for everyone. Here's the [link](https://react-hook-form.com/advanced-usage/#AccessibilityA11y) to the official React Hooks Form docs about accessibility.

edenbekele replied:
This is a great idea! We have altered the error validations to be more accessible.

alexandra-lim reviewed (commented):
This is amazing work @edenbekele and @terchiem! Super cool use of React Hooks Form, I didn't know that was a thing! Just left a few comments regarding the use of it in the code.

GabbyJ reviewed (changes requested):
The AC is all met but we think the resetting data should be addressed. Good job!

edenbekele committed:
Implemented reset on submission of list item

Co-authored-by: Terry Chiem <terrychiem@gmail.com
Co-authored-by: Terry Chiem <terrychiem@gmail.com>
Co-authored-by: Eden Bekele <edbek@yahoo.com>

edenbekele committed:
Made error validations more accessible

> Co-authored-by: Terry Chiem <terrychiem@gmail.com>
> Co-authored-by: Eden Bekele <edbek@yahoo.com>

jmkuchman commented on src/Components/AddItem.jsx:
```diff
@@ -1,5 +1,85 @@
+  const [token, setToken] = useState('test-token');
+  const { register, handleSubmit, reset, errors } = useForm();
+  const onSubmit = (data) => {
+    const currentList = db.collection('lists').doc(token);
```
I like that you're thinking ahead here and saving the lists in terms of their token! 👍

jmkuchman commented on src/Components/AddItem.jsx:
```diff
@@ -1,5 +1,85 @@
+        name="itemName"
+        defaultValue=""
+        aria-invalid={errors.itemName ? 'true' : 'false'}
+        ref={register({ required: true, minLength: 3 })}
```
I'm curious, what made you choose a value of 3 for minimum length?

jmkuchman reviewed (approved):
Wow, really solid work! You've hit all of the AC and built a great foundation for the form moving forward. I like how you utilized React Hook Forms and implemented validation and accessibility.

mxmason commented on src/Components/AddItem.jsx:
```diff
@@ -1,5 +1,85 @@
+    <form onSubmit={handleSubmit(onSubmit)}>
+      <br />
+      <label htmlFor="itemName">Item Name: </label>
+      <input
```
This input needs an in order to be labeled!

terchiem replied:
Done!

mxmason reviewed (changes requested):
Looks good! One change, please!

mxmason commented on src/Components/AddItem.jsx:
```diff
@@ -1,5 +1,85 @@
+      <br />
+      <span style={{ color: 'red' }}>
+        {errors.itemName && errors.itemName.type === 'required' && (
+          <span role="alert">Item name is required.</span>
```
This needs emphasis!
```suggestion
          <span role="alert">Item name is required!</span>
```

terchiem committed:
Fix broken input label

Co-authored-by: Eden Bekele <edbek@yahoo.com>
Co-authored-by: Terry Chiem <teryrchiem@gmail.com>

terchiem committed:
Remove Counter from App

terchiem committed:
Merge from dev

terchiem committed:
Refactor Button to generate a list on Firestore based on token.
